Add flag to force remove long unregistered nodes

BigDarkClown opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

##### Problem:

The Cluster Autoscaler sometimes cannot remove long unregistered nodes. This happens when removing them would leave their node group with fewer than the minimum size required. This wastes resources, as the nodes cannot be used and might be as well deleted.

##### Solution:
- We've added a new option that lets you forcefully remove these unused machines.
- This option only works if your cloud provider allows deleting machines without checking if it violates the minimum size rule.

##### Background:
- This effort started with this code change: https://github.com/kubernetes/autoscaler/pull/6843

#### Does this PR introduce a user-facing change?
```release-note
Change adds a flag --force-delete-long-unregistered-nodes, which if used enables the Cluster Autoscaler to remove the long unregistered nodes even if it would break the min size constraints of their node group.
```

Shubham82 commented:
Hi @BigDarkClown, the mandatory test case `Tests / test-and-verify (pull_request)` is failed, Please address it.

Thanks!

BigDarkClown commented:
> There are the gofmt and spelling errors, due to which this test case failed.
> 
> Run `hack/update-gofmt.sh` and `hack/verify-spelling.sh` to fix this issue.

Thanks @Shubham82, I've updated now. Turns out that gofmt picked up a lot of problems in cloudproviders I've modified.

x13n commented on cluster-autoscaler/core/static_autoscaler.go:
```diff
@@ -785,29 +785,38 @@ func (a *StaticAutoscaler) removeOldUnregisteredNodes(allUnregisteredNodes []clu
-		if len(unregisteredNodesToDelete) > possibleToDelete {
-			klog.Warningf("Capping node group %s unregistered node removal to %d nodes, removing all %d would exceed min size constaint", nodeGroupId, possibleToDelete, len(unregisteredNodesToDelete))
-			unregisteredNodesToDelete = unregisteredNodesToDelete[:possibleToDelete]
+		if !a.ForceDeleteLongUnregisteredNodes {
```
This function is getting pretty long, maybe there's some part that could be extracted with a meaningful name?

BigDarkClown replied:
I've split the part that lists the nodes to be removed into different function + tidied up the comments and removed unused arguments.

x13n commented on cluster-autoscaler/main.go:
```diff
@@ -279,6 +279,7 @@ var (
 	checkCapacityBatchProcessing                 = flag.Bool("check-capacity-batch-processing", false, "Whether to enable batch processing for check capacity requests.")
 	checkCapacityProvisioningRequestMaxBatchSize = flag.Int("check-capacity-provisioning-request-max-batch-size", 10, "Maximum number of provisioning requests to process in a single batch.")
 	checkCapacityProvisioningRequestBatchTimebox = flag.Duration("check-capacity-provisioning-request-batch-timebox", 10*time.Second, "Maximum time to process a batch of provisioning requests.")
+	forceDeleteLongUnregisteredNodes             = flag.Bool("force-delete-unregistered-nodes", false, "Whether to enable force deletion of long unregistered nodes, regardless of the min size of the node group the belong to.")
```
If we control this via flag, it will stay a flag forever and most people won't turn it on. Could this be inferred from cloudprovider support for `ForceDeleteNodes` instead?

BigDarkClown replied:
I thought about this, but I think there will be cases when even though the `ForceDeleteNodes` is implemented, the Cluster Autoscaler user will not want to go below the min size. This is not the case for GCE/GKE, but I cannot guarantee that this will be the case for other providers, so I think keeping this as a flag is a better option.

We can revisit this later, but for now I don't think we should enable it without the possibility of disabline it if anything goes wrong.

x13n replied:
Fair enough, but right now this is opt-in, not opt-out. Perhaps we could set it to `true` by default, call `ForceDeleteNodes` on init and fall back to `false` if the error is `NotImplemented`?

As a side note, I think we could simplify CA a bit if cloudproviders explicitly exposed supported capabilities instead of returning errors from a subset of functions, but that's very out of scope here.

BigDarkClown replied:
That would be a nice workaround for now, but unfortunately it is not possible without a bigger refactor. AutoscalingOptions are initialized before the CloudProvider is created, so we have no way at this point to run `ForceDeleteNodes` to check for `NotImplemented`. We could alternatively do that during execution, but this seems like not-that-great solution.

Let's sync on the idea of exposing supported capabilities on SIG meeting.

x13n commented on cluster-autoscaler/cloudprovider/aws/aws-sdk-go/aws/awserr/error.go:
```diff
@@ -10,24 +10,23 @@ package awserr
-//         }
-//     }
-//
+//	output, err := s3manage.Upload(svc, input, opts)
```
Is this (and all the following) changes to `aws-sdk-go` necessary? Isn't this autogenerated/vendored manually?

BigDarkClown replied:
I was able to remove these changes without triggering the gofmt.

x13n commented:
/lgtm
/approve
